File: year_2023/day_05/day05.py
import logging
import re
from itertools import count

from tools import loader, parsers, timer


class Plants:
    def __init__(self, data: list[list[str]]) -> None:
        self.seeds = list(map(int, re.findall(r'\d+', data[0][0])))
        self.tables = []
        for block in data[1:]:
            _block = []
            for line in block[1:]:
                nums = tuple(map(int, line.split()))
                destination = range(nums[0], nums[0] + nums[2])
                source = range(nums[1], nums[1] + nums[2])
                _block.append((source, destination))
            self.tables.append(_block)

    def part_1(self) -> int:
        """
        >>> print(Plants(parsers.blocks('test.txt')).part_1())
        35"""
        result = float('inf')
        for seed in self.seeds:
            current = seed
            for table in self.tables:
                match = None
                for line in table:
                    if current in line[0]:
                        match = line[1].start + current - line[0].start
                current = match or current
            result = min(result, current)
        return result

    @timer.wrapper
    def part_2(self) -> int:
        """Bruteforce, takes a while.
        >>> print(Plants(parsers.blocks('test.txt')).part_2())
        46"""
        a = list(zip(self.seeds[::2], self.seeds[1::2]))
        seeds = [range(i[0], i[0] + i[1]) for i in a]

        for seed in count(1):
            if seed % 1_000_000 == 0:
                logging.debug(seed)
            current = seed
            for table in self.tables[::-1]:
                match = None
                for line in table:
                    if current in line[1]:
                        match = line[0].start + current - line[1].start
                current = match or current
            if any(current in s for s in seeds):
                return seed
        raise ValueError('Solution not found')


print(Plants(parsers.blocks(loader.get())).part_1())  # 484023871
print(Plants(parsers.blocks(loader.get())).part_2())  # 46294175

File: year_2023/day_05/day05.go
package main

import (
	"aoc/tools"
	"fmt"
	"slices"
	"strings"
)

type mapping struct {
	destStart int
	srcStart  int
	length    int
}

func part1(ranges [][]mapping, seeds []int) int {
	result := make([]int, len(seeds))
	copy(result, seeds)

	for _, block := range ranges {
		for i, seed := range result {
			for _, m := range block {
				if m.srcStart <= seed && m.srcStart+m.length > seed {
					result[i] = m.destStart + seed - m.srcStart
				}
			}
		}
	}

	return slices.Min(result)
}

func part2(ranges [][]mapping, seeds []int) int {
	seed := 0

	for {
		seed++
		current := seed

		for i := len(ranges) - 1; i >= 0; i-- {
			maps := ranges[i]
			match := 0

			for _, m := range maps {
				if m.destStart <= current && m.destStart+m.length > current {
					match = m.srcStart + current - m.destStart
				}
			}

			if match != 0 {
				current = match
			}
		}

		for i := 0; i < len(seeds); i += 2 {
			start, length := seeds[i], seeds[i+1]
			if current > start && current < start+length {
				return seed
			}
		}
	}
}

func solve(input string) (int, int) {
	blocks := tools.SplitBlocks(tools.ReadLines(input))
	seeds := tools.StrToInt(strings.Fields(blocks[0][0])[1:])
	ranges := [][]mapping{}

	for _, block := range blocks[1:] {
		b := []mapping{}

		for _, line := range block[1:] {
			nums := tools.StrToInt(strings.Fields(line))
			b = append(b, mapping{nums[0], nums[1], nums[2]})
		}

		ranges = append(ranges, b)
	}

	return part1(ranges, seeds), part2(ranges, seeds)
}

func main() {
	fmt.Println(solve("test.txt"))
	fmt.Println(solve(tools.GetData(2023, 5)))
}
